modules/dbnd-airflow/src/dbnd_airflow/scheduler/single_dag_run_job.py

```diff
@@ -766,3 +801,45 @@ def heartbeat_callback(self, session=None):
             # self.executor.terminate()
             return
         self.terminating = True
+
+
+@provide_session
+def find_and_kill_zombies(args, session=None):
+    """
+    Find zombie DagRuns and TaskInstances that are in not end state and for which related
+    BackfillJob has failed.
+    """
+    print("Cleaning zombie tasks")
+    failed_jobs_id_qry = (
+        session.query(BaseJob.id)
+        .filter(BaseJob.state == State.FAILED, BaseJob.job_type == "BackfillJob")
+        .all()
+    )
+    failed_jobs_id = set(r[0] for r in failed_jobs_id_qry)
+
+    print("Found {} failed jobs".format(len(failed_jobs_id)))
+
+    running_dag_runs = session.query(DagRun).filter(DagRun.state == State.RUNNING).all()
+
+    for dr in running_dag_runs:
+        if not isinstance(dr.conf, dict):
+            continue
+
+        job_id = dr.conf.get("job_id")
+        if job_id not in failed_jobs_id:
+            continue
+
+        print(
+            "- DagRun {} is in state {} but related job has failed\n".format(
+                dr, dr.state
+            )
+        )
+        # Now the DR is running but job running it has failed
+        dr.state = State.FAILED
+        session.query(TI).filter(
+            TI.dag_id == dr.dag_id,
+            TI.execution_date == dr.execution_date,
+            TI.state != State.SUCCESS,
+        ).update({TI.state: State.FAILED, TI.end_date: timezone.utcnow()})
```

Comment:
should be the same logic as "complete" - we can extract that and reuse it here as well

Reply:
You mean this? I agree that it should be same but I would be in favor in keeping duplication of this fragment - it's much more simpler to understand what's going on in the query. 
```
        qry = session.query(TI).filter(
             TI.dag_id == self.dag_id,
             TI.execution_date == self.execution_date,
             TI.state.notin_(END_STATES),
         )
```

Reply:
almost whole  _validate_dag_run_completed_state can be reused here . let's just make it standalone function and use in both places. 

personally I always in favor of following constructs
  if something_wrong:
        fix_something_wrong 
than 
   _validate_and_fix_if_something_wrong. ( especially if the function called only once)